### PR TITLE
feat(identity): expose `KeyType` for `PublicKey` and `Keypair`

### LINKED ISSUE #3649: `identity::Keypair` ergonomics improvements
## Motivation

After https://github.com/libp2p/rust-libp2p/pull/3350 a couple of things that were easily achievable with a match statement are convoluted/potentially less performant. 

Simple examples:
### Accessing the kind of key by value when supporting more than one kind requires clone

Getting the value of the key while supporting more than one key type
```rust
match key {
    Secp256k1(pk) => {         
        do_smth_with_secp256k1_value(pk)
    }                                     
    Ed25519(pk) => {           
        do_smth_with_ed25519_value(pk)  
    }
}
```
Now this requires a clone since `into_secp256k1` consumes the value
```rust
if let Some(pk) = key.clone().into_secp256k1() {
        do_smth_with_secp256k1_value(pk)
} else if let Some(pk) = key.into_ed25519() {
        do_smth_with_ed25519_value(pk)  
}
```

### Accessing the kind of key by reference

```rust
match &key {
    Secp256k1(pk /* A reference */) => {         
        do_smth_with_secp256k1_ref(pk)
    }                                     
    Ed25519(pk /* A reference */ ) => {           
        do_smth_with_ed25519_ref(pk)  
    }
}
```
Now this requires a clone since `into_secp256k1` consumes the value
```rust
if let Some(pk) = key.clone().into_secp256k1() {
        do_smth_with_secp256k1_ref(&pk)
} else if let Some(pk) = key.into_ed25519() {
        do_smth_with_ed25519_ref(&pk)  
}
```

### Supporting a new key kind
Before: Enable the feature and compiler would tell of a missing match branch. 
Now: Enable feature and dig in the code where the `into_keykind()` are, to add a new one

## Proposed sol

Assuming the code base is going ahead with this deprecation,

for 1: Implement `TryFrom`/`TryInto` for the necessary types. Alternative, add the matching `try_into_secp256k1` and others. In these, the error type should give back the original key to be used.

for 2: Implement `AsRef` for the necessary types. Alternative, add the matching `as_secp245k1()` and others

for 3. Can't see a good solution right now

## Are you planning to do it yourself in a pull request?

Maybe.

### PR DESCRIPTION
**Pull request by gurinderu**
## Description

Resolves https://github.com/libp2p/rust-libp2p/issues/3649.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: identity/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,10 @@
+## 0.2.1 - unreleased
+
+- Add key_type method to public key and key pair.
```
Comment (thomaseizinger) [inline identity/CHANGELOG.md]:
```suggestion
- Expose `KeyType` for `PublicKey` and `Keypair`.
```

**Review by thomaseizinger — DISMISSED** [review]
Awesome, thanks!

**Review by thomaseizinger — APPROVED** [review]

File: identity/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,10 @@
+## 0.2.1 - unreleased
```
Comment (mxinden) [inline identity/CHANGELOG.md:1]: This is missing a version bump in `identity/Cargo.toml`. Without the version bump this patch will not be released any time soon.

Comment (thomaseizinger) [reply identity/CHANGELOG.md:1]: Whoops. My bad for overlooking this.

Comment (thomaseizinger) [reply identity/CHANGELOG.md:1]: Sent follow-up in https://github.com/libp2p/rust-libp2p/pull/4125.

**Pull request merged**